contracts/ERC-721/ERC721PoolFactory.sol

```diff
@@ -39,7 +43,8 @@ contract ERC721PoolFactory is IERC721PoolFactory {
         string memory name = string.concat(IERC721Metadata(asset).name(), " Pool");
         string memory symbol = string.concat(IERC721Metadata(asset).symbol(), "p");
 
-        bytes32 salt = keccak256(abi.encodePacked(asset));
+        bytes32 salt = keccak256(abi.encodePacked(asset, nonce));
+        nonce++;
```

Comment:
@scorpion9979
As I understand it, by initializing the uint256 public nonce variable with an initial value of zero and incrementing it for each subsequent pool creation, we can still compute the pool address off-chain based on the value of the nonce. However, I would like to understand your perspective more clearly to ensure we are on the same page.

Reply:
@surbhiaudichya Knowing what the current nonce is requires reading it onchain. Also, in the code above, the nonce is being incremented for every single pool creation and not just for the same asset. If we could have the nonce be a mapping where the key is the asset address and we increment the count for each asset separately, I think that could be a better approach. Most assets would have a pool with a nonce of 0, unless the rescue function is called and a new pool is deployed with an incremented nonce.

Reply:
Maybe a solution would be to have all pools behind a proxy, and if the `rescueNft` function is ever called, just switch the proxy implementation to a new instance. That way, addresses can be computed off-chain, they never change from a user / factory pov and the `rescueNft` issue is abstracted away. wdyt ?

Reply:
@scorpion9979, your suggestion of using a unique nonce for each asset and incrementing it upon each new pool creation is a good and simple solution. I see how this could efficiently provide a unique identifier for each pool. However, we might face same issue if the `rescueLastNFT` function is invoked multiple times for the same asset, as we would need to read the current nonce on-chain for calculating the pool address. To keep progressing, I've implemented above solution in the latest [commit](https://github.com/hifi-finance/pooled-nft/pull/9/commits/05b023eff92a0f8c76b458ee93ca8ce8a42297ad), but I still think it's worth considering other options.

@hroussille,  Proxy-based solution is an elegant way to handle this situation. it would involve some modifications to our existing contract structure. I am unsure if adding an upgradable proxy aligns with our project's current direction.